### DISCUSSION TITLE
Update the object used in a livewire component

### DISCUSSION THREAD

**Discussion by fornaserofranco** [livewire/livewire#5138] [1. Help] [answered]
Hello everyone. I am developing a shopping cart with livewire. I currently have 2 components: one with a form and one with the cart itself. The form is for adding products to that cart. The problem I'm having is that after I press the "add to cart" button, laravel throws me an error that it can't read the properties of the "products" object. It is as if the object disappeared. I've been investigating and the problem as I understand it occurs because I need to somehow rebuild that object, but I don't know how. It should be in the render method but somehow I have to pass an id to query the server. I'm really confused and I don't know how to solve it. I am going to attach the code so that you better understand what I am talking about.

My livewire component inside blade template
```php
@livewire('form-carrito', ['productos' => $productos])
```

This is the controller of template 

```php
    public function show($id)
        {
            $productos = DB::table('vista_promociones')
                        ->where('agrupa', '=', $id)
                        ->select('precio', 'nombre_cor', 'marca_cor', 'agrupa', 'tamano', 'codigo', 'detalle', 'sin_stock')
                        ->get()
            // bla bla bla
    
            return view('detalle_producto')->with(compact('productos', 'imagenes', 'recomendados'));
        }
```

"form-carrito" (component livewire)
```blade
<div>
    <form wire:submit.prevent="submit(Object.fromEntries(new FormData($event.target)))" class="mb-3">
        @csrf
        @foreach ($productos as $producto)
                <p class="mb-0 fs-12">Código interno: {{$producto->codigo}}</p>
                <div class="input-group is-invalid mb-2 row mx-0">
                    <div class="input-group-prepend col-6 px-0 ">
                        <label class="label_nombre input-group-text w-100 fs-12 fs-xl-14"
                            for="validatedInputGroupSelect">{{ ucwords(strtolower($producto->detalle)) . ' - ' . substr($producto->tamano, 0, 5) }}</label>
                    </div>

                    <select class="custom-select col-4 fs-12 fs-xl-14" name="{{$producto->codigo}}" required >
                        <option value="default">Cantidad</option>
                        @for ($i = 1; $i <= 30; $i++)
                            <option value="{{$i}}">{{$i}}</option>
                        @endfor                                    
                    </select>

                    <div class="input-group-prepend px-0 col-2 ">
                        <label class="input-group-text rounded-right w-100 justify-content-center fs-12 fs-xl-14" >$ {{ number_format($producto->precio, 0, ',', '.') }}</label>
                    </div>
                </div>
          
        @endforeach
        <div class="input-group-append">
            <button class="btn btn-secondary rounded btn-block" id="button_agregar_carrito" type="submit">Agregar al carrito</button>
        </div>
    </form>
</div>
```

Controller of component livewire
```php
class FormCarrito extends Component
{
    public $productos;

    public function render()
    {
        return view('livewire.form-carrito');
    }

    public function submit($formData){
        $codigosProductos = array_column($this->productos->toArray(), 'codigo');   
        Cart::destroy(); 
        sort($codigosProductos);
        $codigosFom = [];    
        foreach ($formData as $codigo => $cantidad) {
            if ($codigo != '_token') {
                $codigosFom[] = $codigo;
            }
        }
        if ($codigosFom == $codigosProductos) {
            foreach ($this->productos as $producto) {
                if ($producto['sin_stock'] == 0 && $formData[$producto['codigo']] != 'default') {
                    Cart::add($producto['codigo'], $producto['nombre_cor'], $formData[$producto['codigo']], $producto['precio']);
                }
            }
        }

        $this->emit('carrito_actualizado');

    }   
}
```

Error after click button form 
![image](https://user-images.githubusercontent.com/108434960/181627219-1783d876-50c1-4aba-80a2-087079737164.png)
![image](https://user-images.githubusercontent.com/108434960/181627274-c264d8ba-e326-4412-9a44-c9b209017925.png)

If I show the elements that have cart with dd(Cart::content()) I can display them correctly. The problem occurs after "emit". The component is re-rendered but does not have the "productos" object to correctly display the form

Any help is welcome, I don't know how to solve this and I've been investigating for hours without progress.

**Comment by joshhanley** [comment]
@fornaserofranco please don't paste screenshots of code, as we can't copy it and modify it to assist you with your issue. Paste code in as text.

When you paste code into github, you need to surround it in triple back ticks and then specify the file type like php or blade.

For example
\```php
\\ PHP code here
\```

\```blade
\\ HTML and blade code here
\```

Thanks!

**Comment by joshhanley** [comment]
@fornaserofranco it looks like the objects you are saving to `$productos` aren't supported by Livewire's serialisation, hence is getting converted to an array.

Can you share how you are getting/quering `$productos`, which you are then passing into your component `@livewire('form-carrito', ['productos' => $productos])`?

Thanks!

**Comment by fornaserofranco** [comment]
@joshhanley Sorry for the screenshots. I will try to edit the post.

```php
$productos = DB::table('vista_promociones')
                    ->where('agrupa', '=', $id)
                    ->select('precio', 'nombre_cor', 'marca_cor', 'agrupa', 'tamano', 'codigo', 'detalle', 'sin_stock')
                    ->get();
```

This is how I get the products. I had already been reading that DB::table type queries are not compatible with the type of object that livewire can handle, I saw in a post that they recommended using toArray() after the get but that didn't work for me.

**Comment by joshhanley** [comment]
@fornaserofranco all good 🙂

Ok your query is definitely the issue. Livewire can only serialise Eloquent Collections. It doesn't support custom queries/selects.

So if you had `Post::where()->get()`, that would return an Eloquent Collection of Post Eloquent Models, which Livewire can work with.

But using `DB` returns a collection of stdObj which Livewire doesn't support.

Calling `toArray()` as suggested would be best. Seems strange that isn't working.

Can you do a `dd($productos)` and screenshot the output so I can see how it is displaying straight from the query?

**Reply by fornaserofranco** [reply]
@joshhanley Yes, no problem. This is what dd($productos) returns

![image](https://user-images.githubusercontent.com/108434960/181668789-988d73bd-4ebc-4cee-9ac2-23679aaa0165.png)

I was just doing tests, I tried again with the toArray() and it still doesn't work. I access the values in the form $array[i]->key. But again, it tells me again that I'm trying to read properties from a 'non-object' after submitting the form. It's like the object I'm passing from blade to livewire component disappears. Maybe it's something I have to do in the render method?

**Reply by Qirel** [reply]
Using `->toArray()` when you pass in the data, so like this, 

```php
@livewire('form-carrito', ['productos' => $productos->toArray()])
```

Then the data is an array, and you'll need to access it with the array syntax, so instead of `$producto->codigo` you'll access it with

```php
$producto['codigo']
```

**Reply by fornaserofranco** [reply]
@Qirel It does not work. If I use the toArray() when I do the query or in the way you mentioned, what I actually get in the livewire component is an array, but if I change the syntax to the form $producto['codigo'] it gives me an error .

This is the error I get if I use the form $producto['codigo']:
![image](https://user-images.githubusercontent.com/108434960/181750436-7f9ebf0c-8677-4d28-99b2-98ac7867e9eb.png)

This is how I receive the data in the component if I do a dd($products) when I do the toArray() in any way: 
![image](https://user-images.githubusercontent.com/108434960/181750719-2a659b9c-c118-4868-af03-0023ebc8145e.png)

**Comment by fornaserofranco** [comment]
Ok... I was doing more tests and I decided to associate "vista_promociones" to a model to make the query in the form VistaPromociones::where(..) and everything works. Now if I have to access the attributes of the array like this: product['code'] when it wouldn't let me before, it doesn't matter where I do toArray(). So why is this happening? Why did you have to use yes or yes in the $producto->key form before, even though it is an array? $productos can only be an Eloquent Collection? Is there no way to continue using the DB::table()? In the current project we do not use models because we have different views on the server that provide us with the data we need, so it does not seem correct to me to make a model for each view. There are many questions but I need to understand why the previous way does not work and now it does.


Controller of livewire component
```php
$productos = VistaPromociones::where('agrupa', '=', $id)
                    ->select('precio', 'nombre_cor', 'marca_cor', 'agrupa', 'tamano', 'codigo', 'detalle', 'sin_stock')
                    ->get();
```

My livewire component inside blade template
```php
@livewire('form-carrito', ['productos' => $productos->toArray()])
```

form-carrito
```blade
<div>
    <form wire:submit.prevent="submit(Object.fromEntries(new FormData($event.target)))" class="mb-3">
        @csrf
        @foreach ($productos as $producto)
                <p class="mb-0 fs-12">Código interno: {{$producto['codigo']}}</p>
                <div class="input-group is-invalid mb-2 row mx-0">
                    <div class="input-group-prepend col-6 px-0 ">
                        <label class="label_nombre input-group-text w-100 fs-12 fs-xl-14"
                            for="validatedInputGroupSelect">{{ ucwords(strtolower($producto['detalle'])) . ' - ' . substr($producto['tamano'], 0, 5) }}</label>
                    </div>

                    <select class="custom-select col-4 fs-12 fs-xl-14" name="{{$producto['codigo']}}" required >
                        <option value="default">Cantidad</option>
                        @for ($i = 1; $i <= 30; $i++)
                            <option value="{{$i}}">{{$i}}</option>
                        @endfor                                    
                    </select>

                    <div class="input-group-prepend px-0 col-2 ">
                        <label class="input-group-text rounded-right w-100 justify-content-center fs-12 fs-xl-14" >$ {{ number_format($producto['precio'], 0, ',', '.') }}</label>
                    </div>
                </div>
          
        @endforeach
        <div class="input-group-append">
            <button class="btn btn-secondary rounded btn-block" id="button_agregar_carrito" type="submit">Agregar al carrito</button>
        </div>
    </form>
</div>
```

**Comment by Martleby** [comment]
@fornaserofranco Here's what I think is happening:

When you first run the query with ->get()->toArray() you are getting an array of standard PHP Objects and so on that initial page load everything works ok when you do ```$producto->codigo```.

However, when you press the submit button, Livewire re-hydrates your $producto property as an array, but the objects are de-serialized as arrays, so on the next page render these are no longer objects but arrays and so ```$producto->codigo``` fails as it isn't an object anymore.

So on the first render of the page you're dealing with an array of objects, but on the second and subsequent renders you're dealing with an array of arrays!

My usual solution to this is to do something like this on the component:

```php
public function hydrateCodigos($value)
{
     $this->codigos = collect($value)->map(fn ($producto) => (object) $producto)->toArray();
}
```

Basically what this does is turn the array of arrays back in to an array of objects whenever Livewire re-hydrates the property.

**Comment by joshhanley** [comment, marked as answer]
@fornaserofranco yeah @Martleby is correct here, `toArray()` on a collection of standard objects will only convert the collection class to an array but not the individual objects. But when Livewire serialises the standard objects, it converts them to arrays, but has no way to know to convert them back.

So there are a couple of solutions:

Do as @Martleby suggested above and use a hydrate hook to convert the arrays back to standard objects, that way you can use the object notation in your view `$producto->codigo`.

Or if you want to keep it simple, convert the objects to arrays before converting the collection to an array, and then use array syntax for everything. That saves you needing to use hydrate hooks everywhere.

Do to this, just map through them:

```php
$productos = DB::table('vista_promociones')
                    ->where('agrupa', '=', $id)
                    ->select('precio', 'nombre_cor', 'marca_cor', 'agrupa', 'tamano', 'codigo', 'detalle', 'sin_stock')
                    ->get()
                    ->map(function($item) {
                        return (array) $item;
                    })
                    ->toArray();
```

That way it is just an array of arrays that you pass around and use for everything.

In regards to your comment about making models, while that will work, as you said that is going to be a lot of work. So I wouldn't suggest doing it. I would just keep using the DB queries and make use of one of the solutions above.

Hope this helps! 🙂

**Reply by fornaserofranco** [reply]
Thank you! I didn't answer any more because I solved the original problem by making a model for that sql view, but I got the same problem again and used your solution. Everything seems to work!

**Reply by joshhanley** [reply]
@fornaserofranco no worries! 🙂
